runtime/memprof.c

```diff
@@ -42,45 +550,612 @@ struct memprof_thread_s {
+      /* An offset entry, which we ought to evict. We'll never be able
+       * to connect this entry to its allocated block, so ideally no
+       * callbacks are run for it. However, some callbacks may already
+       * have been run. If the allocation callback has been run, but
+       * not the deallocation callback, the best we can do is probably
+       * to fake deallocating the block, so that alloc/dealloc
+       * callback counts correspond. */
+      if ((es->t[i].callbacks & CB_MASK(CB_ALLOC)) &&
+          ((es->t[i].callbacks & CB_MASK(CB_DEALLOC)) == 0)) {
+        es->t[i].offset = false;
+        es->t[i].block = Val_unit;
+        es->t[i].deallocated = 1;
+        if (i < es->next) es->next = i;
+      } else {
+        dead = true;
+      }
+    }
+
+    if (dead) {
+      /* entry is evicted, but what if a thread is running it?  This
+       * can happen specifically if an allocation callback stops a
+       * profile and starts a new one: all 'offset' entries are evicted.
+       */
+      memprof_thread_t runner = es->t[i].running;
+      if (runner != NULL) {
+        CAMLassert(runner->running_table == es);
+        CAMLassert(runner->running_index == i);
+        runner->running_table = NULL;
+      }
+    } else { /* preserve this entry */
+      if (i != j) {
+        es->t[j] = es->t[i];
+        /* if a callback is currently being run on this entry,
+         * make sure its index is updated */
+        memprof_thread_t runner = es->t[i].running;
+        if (runner != NULL) {
+          CAMLassert(runner->running_table == es);
+          CAMLassert(runner->running_index == i);
+          runner->running_index = j;
+        }
+      }
+      ++ j;
+    }
+    ++ i;
+    if (es->young == i) es->young = j;
+    if (es->next == i) es->next = j;
+  }
+  es->evict = es->live = j;
+  CAMLassert(es->next <= es->live);
+  CAMLassert(es->young <= es->live);
+
+  ensure_entries(es, 0);
+}
+
+/* Transfer all entries from one entries table to another. */
+
+static void entries_transfer(entries_t from, entries_t to)
+{
+  size_t offset = to->live;
+
+  if (from->live == 0) {
+    return;
+  }
+
+  CAMLassert(to->config == from->config);
+
+  ensure_entries(to, from->live);
+  to->live += from->live;
+
+  for (size_t j = 0; j < from->live; ++j) {
+    to->t[offset + j] = from->t[j];
+    memprof_thread_t thread = from->t[j].running;
+    if (thread) { /* unusual */
+      CAMLassert(thread->running_table == from);
+      CAMLassert(thread->running_index == j);
+      thread->running_table = to;
+      thread->running_index = offset+j;
+    }
+  }
+
+  if (to->young == offset) {
+    to->young = offset + from->young;
+  }
+  if (to->evict == offset) {
+    to->evict = offset + from->evict;
+  }
+  if (to->next == offset) {
+    to->next = offset + from->next;
+  }
+  /* Reset `from` to empty, and allow it to shrink */
+  ensure_entries(from, -from->live);
+  from->young = from->evict = from->next = from->live = 0;
```

Comment:
```suggestion
  from->young = from->evict = from->next = from->live = 0;
  ensure_entries(from, 0);
```

Frankly, using `ensure_entries` in such a way that `live > size` after the call is really misleading, even if it appears that the code is correct.

Reply:
This is good, and also lets me avoid change a signed type to unsigned (the `grow` parameter to `ensure_entries`).